[1단계 - 장바구니] 제나(이정은) 미션 제출합니다.

JeLee-river opened the pull request:
안녕하세요 쵸파! 이번 장바구니 미션의 리뷰이가 된 제나입니다! 2주 간 잘부탁드립니다. ㅎㅎ

---

## 📦 장바구니 미션

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계

- 클라이언트 상태를 효과적으로 모델링하고 관리할 수 있다.
- Jest, React Testing Library(RTL)를 활용하여 주요 기능에 대한 테스트를 작성할 수 있다.

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] RTL 테스트 케이스를 모두 작성했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?

  - 배포 링크 기입: [**장바구니**](https://react-shopping-cart-taupe-gamma.vercel.app/)

- [ ] 리뷰어가 장바구니 추가를 쉽게 할 수 있도록 Curl 명령어를 DM 으로 전달해주세요.

  ```
  curl -X 'POST' \
  '<url>' \
  -H 'accept: */*' \
  -H 'Authorization: Basic <토큰>' \
  -H 'Content-Type: application/json' \
  -d '{
  "productId": 1,
  "quantity": 1
  }'
  ```

### 컴포넌트 분리

![image (1)](https://github.com/user-attachments/assets/365f4c61-7533-47b9-80f9-f39eab32b517)

<br>

### 리뷰 요청 & 논의하고 싶은 내용

### 1) 상태 설계 의도

### 1. 체크 박스 상태 관리

체크 박스의 선택 여부(selected)를 관리하기 위해, cartItem에서 cardId를 추출하여 따로 관리하는 방식을 선택했습니다.

이 과정에서 selected 상태를 서버에서 받아온 cartItem 데이터에 포함시킬지, 포함시키지 않을지 고민했으나 페어와 장단점을 정리한 끝에 포함하지 않기로 결정했습니다. 저희가 생각한 장단점은 다음과 같습니다.

**cartItem 데이터에 포함**
  - 장점
      - id를 이용하여 선택된 장바구니 데이터를 찾는 추가적인 계산이 필요없다.
  - 단점
      - 서버 데이터와 동기화되지 않는다.
          - 서버에서 받는 데이터랑 클라이언트에서 관리하는 상태가 달라진다.

<br>

**cartItem 데이터에 미포함**
  - 장점
      - **서버 데이터와 클라이언트 상태를 명확히 분리할 수 있다.**
      - id만 따로 관리하기 때문에, 관리하기 편하다.
      - 서버 데이터와 동기화 문제가 발생하지 않는다.
  - 단점
      - id를 이용하여 선택된 장바구니 데이터를 찾는 추가적인 계산이 필요하다.
      - 자식 컴포넌트에 상태를 내려줄 때 selected 상태를 따로 계산해서 내려줘야 함.

이에 따라, selected 상태는 클라이언트에서만 사용하는 UI의 상태이기 때문에 carItem 데이터와 분리해서 관리하는 것이 데이터 일치, 확장성 측면에서 좋다고 생각하여, 이와 같이 상태를 설정했습니다.

<br>

### 2. 장바구니 아이템 수량(quantity)의 상태 관리

서버에서 받아오는 cartItem 데이터에서 장바구니 아이템 수량(quantity)를 별도의 상태로 관리하도록 했습니다.

이와 같이 설계한 이유는 **수량 증감으로 변경되는 quantity에 낙관적 업데이트를 적용하기 위해서입니다.**

그러나 quantity를 상태로 관리했을때, cartItem에 의존하는 파생 상태가 되면서 SSOT 원칙에 어긋난 코드가 되었습니다. 

이에 대해 다른 방법이 없을지 페어와 고민했으나 대안을 찾지 못했고, SSOT 원칙을 지키는 것보다는 UX적으로 낙관적 업데이트를 지원하는 것이 더 좋을 것 같다고 판단하여 quantity를 상태로 관리하게 되었습니다.

아래는 저희 페어가 정리한 상태 관리 여부에 따른 장단점 비교입니다.

**[상태로 관리]**
  - 장점
      - 낙관적 업데이트가 된다. 네트워크가 느리더라도 사용자가에게 빠른 피드백 가능
  - 단점
      - SSOT를 지키지 못한다. 진실의 원천이 2개가 됨 (카트 상태랑 커스텀 훅에 있는 quantity 상태 2개가 관리됨)

**[상태로 관리 x]**
  - 장점 : SSOT를 지킬 수 있다.
  - 단점 : 네트워크가 느린 경우 사용자 경험에 부정적인 영향을 미친다.

<br>

### 3. useCheckboxHandler 커스텀 훅 내부의 useRef

체크 박스로 선택한 아이템들의 초기값을 서버 데이터로 설정하기 위해 useRef를 사용하였습니다.

현재, 체크 박스의 체크 여부를 각 장바구니 아이템의 id 배열(selectedCartId)로 상태 관리하고 있습니다. 

요구 사항에 따라, 처음 장바구니 페이지에 접속했을때 모든 아이템들이 체크가 되도록 설정해주어야 했기 때문에, 서버에서 cartItems를 패칭한 후, 여기서 id를 추출하여 useEffect로 다시 selectedCartId의 상태를 업데이트 하도록 했습니다.

그러나 이 id는 cartItems 상태에 의존하기 때문에, 수량 조절, 삭제 등 체크 박스와 관련없는 기능이 실행될 때에도 cartItems의 변경에 의해 체크 여부도 강제로 변경되는 문제가 발생했습니다.

이를 해결하고자 방법을 고민했으나, useRef로 렌더링 횟수를 관리하여, 데이터 패칭 시점까지만 useEffect에 의한 selectedCartId의 변경이 가능하도록 하는 방법 외에는 다른 아이디어를 생각하지 못하여 이와 같이 구현했습니다.

<br>

### 2) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

이번 단계에서 가장 많이 고민했던 문제는 장바구니 아이템 수량(quantity)를 상태로 관리할지의 여부였습니다. 

코드 의도에서 말씀드린 것처럼 UX적으로 낙관적 업데이트가 주는 이점이 크다고 느꼈기 때문에 SSOT에 어긋나지만, quantity를 별도의 상태로 관리하기로 했습니다.

<br>

### 3) 이번 리뷰를 통해 논의하고 싶은 부분

### 1. 파생 상태를 생성하는 기준

파생상태를 생성하는 범용적인 기준이 있을지 궁금합니다.

레벨2 초반에 React에 대해 공부하면서, 데이터 불일치 문제와 유지보수 측면에서의 어려움 때문에 파생상태를 지양해야 한다는 점을 학습했습니다.

그러나 이번 미션에서 낙관적 업데이트를 위해 불가피하게 파생 상태를 생성해야 했을 때 SSOT 원칙에 어긋나더라도 파생 상태를 만드는 것이 최선인지 고민이 되었습니다.

이에 대한 쵸파의 의견과, 파생 상태를 생성하는 기준이 있다면 무엇일지 여쭙고 싶습니다!


---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 클라이언트 상태관리

- 원본상태/파생상태를 적절히 구분하여 선언하였나요?
- React state 를 불필요하게 선언한 부분은 없나요?
- 상태 관리 로직의 책임이 적절히 응집/분리되었나요? (ex. reducer, hook, Context)

### 2. MSW/Test

- 주요 기능을 적절히 정의하였나요?
- 주요 기능/예외에 대한 테스트가 충분히 이루어졌나요?

woowa-euijinkk committed:
chore: 의존성 제거

yeongipark committed:
chore: emotion 설정

JeLee-river committed:
docs: 기능 명세서 작성

yeongipark committed:
style: 전역 스타일 설정

yeongipark committed:
feat: Header 컴포넌트 구현

yeongipark committed:
style: 전역 스타일 수정

JeLee-river committed:
feat: Description 컴포넌트 구현

JeLee-river committed:
style: Header fixed 설정

yeongipark committed:
feat: QuantityControlButton 컴포넌트 구현

yeongipark committed:
feat: CheckBox 컴포넌트 구현

yeongipark committed:
feat: CartItem 컴포넌트 구현

JeLee-river committed:
feat: Receipt 컴포넌트 구현

yeongipark committed:
feat: SubmitButton 컴포넌트 구현

JeLee-river committed:
feat: Cart 페이지 구현

JeLee-river committed:
feat: 장바구니 아이템이 없는 경우 메시지 렌더링

yeongipark committed:
feat: OrderSummary 페이지 구현

JeLee-river committed:
feat: Checkbox 컴포넌트 구현

JeLee-river committed:
feat: apiClient 구현

JeLee-river committed:
feat: 데이터를 받아오는 useFetch 훅 구현

JeLee-river committed:
feat: 장바구니 아이템 api 연결

JeLee-river committed:
feat: 장바구니 아이템 수량 조절 기능 추가

yeongipark committed:
feat: 장바구니 아이템 삭제 기능 추가

JeLee-river committed:
feat: Checkbox 로직 구현

yeongipark committed:
feat: 주문 금액 계산 로직 구현

yeongipark committed:
feat: 주문 확인 버튼 로직 구현

JeLee-river committed:
feat: router 설정

JeLee-river committed:
feat: 주문 확인 페이지 구현

yeongipark committed:
feat: 헤더 아이콘 라우터 연결

JeLee-river committed:
docs: 기능 명세서 업데이트

JeLee-river committed:
style: 장바구니 페이지 스타일 개선

yeongipark committed:
style: 페이지 스타일 수정

JeLee-river committed:
feat: fallback 이미지 설정

yeongipark committed:
style: 페이지 스타일 수정

yeongipark committed:
fix: 수량 조절 버튼 눌렀을 때 전체 선택 되는 문제 해결

yeongipark committed:
docs: 테스트 명세 작성

JeLee-river committed:
test: CheckBox 컴포넌트 테스트 추가

woowa-euijinkk committed:
docs: PR template

yeongipark committed:
test: 전체선택 버튼 테스트

yeongipark committed:
chore: npm run test 명령어 변경

yeongipark committed:
test: any타입 FetchDataProps로 변경

yeongipark committed:
test: 주문확인 버튼 테스트 작성

yeongipark committed:
chore: 주석 제거

yeongipark committed:
docs: 테스트 목록 업데이트

yeongipark committed:
chore: 테스트 파일 이름 변경

JeLee-river committed:
Merge branch 'main' into step1

JeLee-river committed:
chore: import 문 순서 조정

JeLee-river committed:
fix: 주문확인 페이지의 헤더에 뒤로가기 아이콘이 나오지 않는 현상 수정

JeLee-river committed:
refactor: 배포 페이지에 아이콘이 뜨도록 경로 수정

bassyu commented on src/hooks/useFetch.ts:
범용적으로 추상화 잘 해주셨네요~! 👍 

useMutation이랑 거의 비슷한 것 같은데, 하나로 합치거나, useFetch만의 특징이나 기능을 더 살려주면 어떨까요?

JeLee-river replied:
사실 useFetch와 useMutation을 분리할지, 하나로 합칠지 페어와 고민을 했습니다. 당시에는 하나로 합치면 API 호출 결과의 타입을 명확히 지정하기 어렵고, onSuccess 내부에서 반환값이 있는 경우와 없는 경우를 모두 처리해야 해서 이해하기 어려운 구조가 될 수 있다는 점을 고려해 두 개로 분리했었습니다.

그런데 리뷰를 반영하면서, onSuccess의 input을 옵셔널로 두고, 이와 상관없이 onSuccess가 호출되게 한다면 충분히 단순한 구조로 하나의 훅으로 합칠 수 있겠다는 생각이 들었습니다. 따라서 하나로 합치는 방향으로 바꿔보겠습니다.

JeLee-river replied:
useFetch와 useMutation를 하나로 합쳤습니다!

[refactor: useFetch와 useMutation으로 분리된 훅을 하나로 통합 (5618edc)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/5618edcbc1dcef707abe0e2c061175911161b385)

bassyu replied:
넵 합치는 방향도 좋네요~!

bassyu commented on src/hooks/useFetch.ts:
```diff
@@ -0,0 +1,32 @@
+  const [isLoading, setIsLoading] = useState(false);
+
+  const fetchData = useCallback(
+    async ({ apiCall, onSuccess, onError }: FetchDataProps<T>) => {
```
fetchData 마다 apiCall을 다르게 받을 수 있네요!
지금은 여러개의 api 호출을 하나의 isLoading 상태로 관리하고 있는데요,
각 API마다 각자의 상태를 가지는 방향은 어떨까요?

JeLee-river replied:
여러 패칭 로직이 동시에 수행되면 isLoading이 공유될 위험이 있으니, 좋은 것 같습니다!!

JeLee-river replied:
api 별로 로딩 상태를 관리할 수 있도록 로딩 상태를 Map 객체로 수정했습니다! 그리고 각 key string는 api로 받아오거나 수정하는 데이터의 상태 명으로 설정했습니다. (boolean → Map<FetchKeyType, boolean>)

[refactor: api 별로 loading 상태를 관리할 수 있도록 Map 객체로 상태를 변경 (9927b4f)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/9927b4fdb55924e2c1e054718d694ce10db2e2fc)

[refactor: Loading Map 객체에 접근할 수 있는 key의 명시적인 타입을 지정 (29c213d)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/29c213da51c7ea0b5cb0fe77648c1b77bc947b6e)

[refactor: useFetch의 반환값 isLoading이 key에 맞는 로딩 상태를 반환하도록 수정 (4948aa3)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/4948aa3d221abf740923682546b069cf073eac8d)

bassyu replied:
좋습니다! 공유되는 것이 목적일 수도 있지만, 이게 혼란을 주거나 불편할 수도 있겠네요!

다만 1개의 훅당 1개의 key와 loading 상태를 관리하는 것으로 보이는데, boolean 대신 Map으로 해주신 이유가 궁금합니다!

bassyu commented on src/pages/Cart/Cart.tsx:
```diff
@@ -0,0 +1,101 @@
+    toggleSelect,
+    isAllSelected,
+    isSelected,
+  } = useCheckboxHandler(cartItems);
```
> 이에 따라, selected 상태는 클라이언트에서만 사용하는 UI의 상태이기 때문에 carItem 데이터와 분리해서 관리하는 것이 데이터 일치, 확장성 측면에서 좋다고 생각하여, 이와 같이 상태를 설정했습니다.

장단점 잘 고민 해주셨네요! 저도 제나 의견에 공감됩니다~!

제 개인적인 경험에서, 서버 상태와 클라이언트 상태를 구분했을 때의 장점이 꽤 컸습니다!

bassyu commented on src/components/CartItem/CartItem.tsx:
```diff
@@ -0,0 +1,78 @@
+  toggleSelect,
+}: CartItemProps) {
+  const { id: cartId, product, quantity: initialQuantity } = cartItem;
+  const { decreaseQuantity, increaseQuantity, deleteCartItem, quantity } =
```
> 이에 대해 다른 방법이 없을지 페어와 고민했으나 대안을 찾지 못했고, SSOT 원칙을 지키는 것보다는 UX적으로 낙관적 업데이트를 지원하는 것이 더 좋을 것 같다고 판단하여 quantity를 상태로 관리하게 되었습니다.

> 그러나 이번 미션에서 낙관적 업데이트를 위해 불가피하게 파생 상태를 생성해야 했을 때 SSOT 원칙에 어긋나더라도 파생 상태를 만드는 것이 최선인지 고민이 되었습니다.


UX를 우선적으로 생각해주신 점이 좋네요~! 👍 👍 

이미 느끼신 것처럼 아쉬운 점은, 같은 데이터를 의미하는 state 여러개가 동시에 존재한다는 점인데요!
이때 예측되는 이슈가 있을까요?

그래서 해결해 보자면, 낙관적 업데이트를 적용하는 동시에 state는 1개로 공유하면 어떨까요?
다시 말해보자면 원본 state를 활용해서 새로운 state를 만드는 대신, 원본 state를 그냥 수정해버리는 방향은 어떨까요?

JeLee-river replied:
1. 같은 데이터를 의미하는 state가 여러개 존재하기 때문에 서버측에서 데이터를 받아올 때마다 관련 state를 모두 업데이트 해줘야하는 문제가 있습니다. 또한 같은 데이터에서 파생된 state들을 서로 동기화 해주지 않으면 데이터 불일치 문제가 발생할 수 있습니다!

2. 오 낙관적 업데이트를 위해 새로운 state를 만들어야 한다고만 생각했는데, 원본을 수정하는 방향도 생각해볼 수 있군요..! 코멘트 주신 내용을 참고하여 어떻게 바꿀 수 있을지 고민해보겠습니다!! 👍

JeLee-river replied:
cartItems의 파생 상태인 quantity를 제거하고, 낙관적 업데이트 로직이 cartItems를 바로 수정하도록 리팩토링했습니다!

[refactor: 파생 상태 대신 원본 상태를 사용하도록 개선 (a9dccce)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/a9dcccef253b5d9de90ec56b9fa6d7b2931b18d1)

bassyu replied:
정리 잘 해주셨네요! 👍 👍

bassyu commented on src/hooks/useCheckboxHandler.ts:
```diff
@@ -0,0 +1,50 @@
+
+const useCheckboxHandler = (cartItems: CartItemType[]) => {
+  const allCartIds = useMemo(
+    () => [...cartItems].map((item) => item.id),
```
메모이제이션 해주신 이유가 궁금합니다~!

JeLee-river replied:
현재 코드는 selectedCartIds의 초기값으로 전체 id(allCartIds)를 설정하기 위해 useEffect를 사용하고 있습니다. 이때, useEffect 내부 로직을 'allCartIds가 빈 배열에서 실제 id 배열로 처음 바뀌는 시점'까지만 실행되도록 하고 싶어, 특정 렌더링 횟수 이하일 때만 상태를 변경하도록 조건문을 추가했습니다. 

하지만 allCartIds를 매 렌더링마다 생성하고 있기 때문에 참조가 항상 달라져서 useEffect가 계속 실행되었고, 결과적으로 렌더링 횟수가 의도보다 빠르게 늘어나 내부 로직 실행 시점을 잘못 판단하는 문제가 발생했습니다.

따라서 allCartIds가 실제로 변경되는 시점인, cartItems가 변경될 때에만 참조가 바뀌도록 메모이제이션하여, 의도한 타이밍까지만 useEffect가 실행되도록 하고자 했습니다!

하지만 selectedCartIds의 초기값 설정 로직을 수정한다면, useMemo를 사용하지 않아도 될것 같습니다. 😂

JeLee-river replied:
selectedCartIds의 초기값으로 전체 id를 설정하는 로직을 개선하면서, 불필요해진 useMemo를 제거했습니다!!

[refactor: 초기 화면에서 전체 선택이 체크되는 로직을 개선 (73ebf0d)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/73ebf0d969fd813123f9c6c5b20493a276b70713)

bassyu replied:
기존에 deps로 사용하고 있었네요!

bassyu commented on src/hooks/useCheckboxHandler.ts:
```diff
@@ -0,0 +1,50 @@
+
+const useCheckboxHandler = (cartItems: CartItemType[]) => {
+  const allCartIds = useMemo(
+    () => [...cartItems].map((item) => item.id),
```
map은 새로운 객체를 반환해서 그대로 사용해도 괜찮겠네요!
```suggestion
    () => cartItems.map((item) => item.id),
```

JeLee-river replied:
selectedCartIds의 초기값으로 전체 id를 설정하는 로직을 개선하면서, cartItems로 allCartIds 계산을 바로 수행하도록 변경했습니다!

[refactor: 초기 화면에서 전체 선택이 체크되는 로직을 개선 (73ebf0d)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/73ebf0d969fd813123f9c6c5b20493a276b70713)

bassyu commented on src/hooks/useQuantityControl.ts:
```diff
@@ -0,0 +1,66 @@
+    await mutateData({
+      apiCall: () => increaseCartItem(cartId, quantity),
+      onSuccess: refetchCartItem,
+      onError: () => {
+        setQuantity((prev) => prev - 1);
+      },
```
낙관적 업데이트 후 에러 상황까지 고려 해주셨네요! 💯

bassyu commented on src/components/CartItem/CartItem.tsx:
전체 선택을 하고 아이템 하나를 삭제하면, 전체 선택이 풀리는 이슈가 있습니다!

아이템과 체크박스 상태의 sync가 맞는지 테스트로 추가해봐도 괜찮겠네요~!

JeLee-river replied:
전체 선택 로직을 개선하면서 문제를 해결했습니다! 추가로, 개별 아이템을 삭제했을 때 전체선택 체크박스 상태의 동작을 테스트하는 코드를 추가했습니다!


[refactor: 초기 화면에서 전체 선택이 체크되는 로직을 개선 (73ebf0d)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/73ebf0d969fd813123f9c6c5b20493a276b70713)

[test: 아이템이 삭제된 경우의 선택 로직 테스트 추가 (9b25022)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/9b2502234c1ca5b9b46c4e41bd89d73bd9022538)

bassyu commented on src/hooks/useCheckboxHandler.ts:
```diff
@@ -0,0 +1,50 @@
+    () => [...cartItems].map((item) => item.id),
+    [cartItems]
+  );
+  const initialRef = useRef(0);
+  const [selectedCartIds, setSelectedCartIds] = useState<number[]>([]);
+
+  useEffect(() => {
+    if (initialRef.current < 3) {
+      setSelectedCartIds(allCartIds);
+      initialRef.current += 1;
+    }
+  }, [allCartIds]);
```
> 이를 해결하고자 방법을 고민했으나, useRef로 렌더링 횟수를 관리하여, 데이터 패칭 시점까지만 useEffect에 의한 selectedCartId의 변경이 가능하도록 하는 방법 외에는 다른 아이디어를 생각하지 못하여 이와 같이 구현했습니다.

렌더링 횟수에 의존하는 부분은 깨질 확률이 너무 높아보입니다!

`cartItems` fetch 후에 바로 `setSelectedCartIds`를 해주면 해결 가능합니다!

JeLee-river replied:
cartItems fetch 후에 바로 useEffect로 상태를 변경하도록 하여, 구조를 개선했습니다!

[refactor: 초기 화면에서 전체 선택이 체크되는 로직을 개선 (73ebf0d)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/73ebf0d969fd813123f9c6c5b20493a276b70713)

bassyu replied:
`cartItems` 상태가 업데이트 된 후에 effect가 한번 더 실행되겠네요~!
fetchData의 onSuccess를 활용해보면 어떨까요?

JeLee-river replied:
오 onSuccess를 바로 이용할 수도 있겠군요..! 한 번 바꿔보겠습니다!

bassyu commented on test/AllCheckBox.test.tsx:
```diff
@@ -0,0 +1,97 @@
+    await screen.findByText("전체선택");
+  });
+
+  it("초기 렌더링 시 전체선택 버튼과 모든 아이템 체크박스가 체크된 상태여야 한다", async () => {
```
UI에 의존하지 않고, 오직 훅의 상태 데이터만 활용해서 체크박스를 테스트 해보면 어떨까요?
물론 지금 테스트도 의미가 있기 때문에 둘 다 있으면 좋을 것 같아요!

JeLee-river replied:
장바구니 아이템을 선택하는 로직 자체를 테스트하는 코드를 추가했습니다!

[test: 장바구니 아이템 선택 로직을 테스트하는 코드 추가 (4dce546)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/4dce54684ec12bdf96702d08e06f91cbb8312c1d)

bassyu reviewed (changes requested):
안녕하세요 제나! 만나서 반갑습니다!
이번 미션 잘 부탁드립니다~! 🙇 

PR에 주로 고민 했던 주제의 장단점을 잘 정리해주셨네요! 
이런식으로 장단점을 고민하며 개발하는 것이 좋다고 생각합니다~! 👍 👍 

기능도 잘 동작하고, 컴포넌트도 적절하게 잘 분리해주셨네요!
질문 주신 effect 부분은 위험해 보여서 코멘트 드렸습니다~!

---

1단계 수고 많으셨습니다!
코멘트 확인 후 다시 요청주세요~!

JeLee-river committed:
refactor: 초기 화면에서 전체 선택이 체크되는 로직을 개선

- 불필요한 useRef를 제거하고 초기 화면에서 전체 선택이 체크되도록 로직 개선

JeLee-river committed:
refactor: 파생 상태 대신 원본 상태를 사용하도록 개선

- quantity를 상태로 관리하지 않고, cartItems 상태를 직접 변경하도록 리팩토링

JeLee-river committed:
refactor: cartItems를 변경하는 함수를 최상단 컴포넌트에서 관리하도록 수정

JeLee-river committed:
refactor: api 별로 loading 상태를 관리할 수 있도록 Map 객체로 상태를 변경

JeLee-river committed:
refactor: cartItem 내부에서 관리하는 핸들러 로직을 최상위 컴포넌트(Cart)로 이동

- 한 곳에서 데이터의 상태를 변경하도록 하기 위해 수정함

- 개별 cartItem 정보가 필요한 핸들러 함수는 cartItems에서 id를 input으로 받아, 아이템을 찾은 뒤 이를 이용해 상태를 변경할 수 있도록 수정

JeLee-river committed:
test: 장바구니 아이템 선택 로직을 테스트하는 코드 추가

JeLee-river committed:
test: 아이템이 삭제된 경우의 선택 로직 테스트 추가

JeLee-river committed:
refactor: useFetch와 useMutation으로 분리된 훅을 하나로 통합

JeLee-river committed:
refactor: Loading Map 객체에 접근할 수 있는 key의 명시적인 타입을 지정

JeLee-river committed:
chore: 불필요한 주석 제거

JeLee-river committed:
refactor: useFetch의 반환값 isLoading이 key에 맞는 로딩 상태를 반환하도록 수정

JeLee-river commented:
안녕하세요 쵸파! 코멘트 주신 내용을 토대로 리뷰 반영을 완료하여 재요청 드립니다!! 아래의 글은 변경 사항을 정리한 내용과 추가 질문입니다!

---


## 주요 변경 사항

### 1️⃣ 장바구니 아이템 선택 로직 개선

**1. 첫 화면에서 전체선택 되도록 하는 로직을 개선함**

- 불필요한 Ref와 메모이제이션을 제거하고, useEffect 내부에서 전체 id를 추출하여 selectedCartIds의 상태를 변경하도록 개선 [(73ebf0d)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/73ebf0d969fd813123f9c6c5b20493a276b70713)

<br>

### 2️⃣ 장바구니 아이템 수량 핸들링 로직 개선

**1. 파생 상태 `quantity` 제거**

- 파생 상태인 quantity를 제거하고, 원본 상태인 cartItems의 수량을 바로 변경하도록 변경 [(a9dccce)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/a9dcccef253b5d9de90ec56b9fa6d7b2931b18d1)

<br>

**2. 수량 변경 핸들러 함수의 위치 변경**

- useQuantityControl을 호출하는 로직을 CartItem 컴포넌트에서 최상위 컴포넌트 Cart로 이동시킴 [(0b5b3c8)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/0b5b3c8fd8ae720fa384e18c62f4b18e9098f29c)

- quantity 대신 cartItems를 직접 전달하여 useQuantityControl의 수량 핸들러 내부에서 id를 받아, 타겟 아이템을 찾고 수량을 변경하도록 리팩토링함

- 핸들러 함수를 최상위 컴포넌트에서 정의하고 props로 내려주기 위해 이와 같이 리팩토링함

<br>

### 3️⃣ 로딩 상태 리팩토링

**1. 로딩 상태의 데이터 타입 변경**

- 로딩 상태를 api 별로 관리하기 위해, useFetch의 isLoading 상태를 기존의 boolean에서 Map<key, boolean> 형식으로 변경함  [(9927b4f)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/9927b4fdb55924e2c1e054718d694ce10db2e2fc)

<br>

**2. 로딩 상태에 접근하는 key 개선**

- Map 객체에 접근하는 key에 명시적인 타입을 부여함 [(29c213d)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/29c213da51c7ea0b5cb0fe77648c1b77bc947b6e)

- useFetch가 전달받은 key에 대응하는 로딩 상태를 반환하도록 리팩토링 [(4948aa3)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/4948aa3d221abf740923682546b069cf073eac8d)

<br>

### 4️⃣ 패칭 인터페이스를 하나로 통합

**1. useFetch와 useMutation을 하나의 인터페이스로 통합**

- 두 개로 분리되어 있던 인터페이스를 하나로 통합함 [(5618edc)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/5618edcbc1dcef707abe0e2c061175911161b385)

<br>

### 5️⃣ 테스트 코드 추가

**1. 장바구니 아이템 선택 로직의 테스트 코드 추가**

- 커스텀 훅 useQuantityControl의 테스트 코드를 추가하여, 체크 박스를 선택/해제하는 기능을 테스트하도록 함 [(4dce546)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/4dce54684ec12bdf96702d08e06f91cbb8312c1d)

<br>

**2. 장바구니 아이템 삭제 테스트 코드 추가**

- 장바구니 아이템을 삭제했을 때, 전체선택과 개별선택이 정상적으로 작동하는지 테스트하는 코드를 추가함 [(9b25022)](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/9b2502234c1ca5b9b46c4e41bd89d73bd9022538)

<br>
<br>

## 추가 질문

### useQuantityControl의 호출 위치와 props

리팩토링 과정에서 useQuantityControl 훅을 호출해 핸들러를 생성하는 로직이 하위 컴포넌트인 CartItem에 위치한 점에 불편함을 느꼈습니다. 전체 데이터를 패칭하는 로직과 cartItems의 상태 정의는 최상위 컴포넌트인 Cart에서 수행하고 있는데, PATCH api 호출과 상태 변경을 수행하는 핸들러 함수는 CartItem에서 생성되고 있어, **상태 관련 로직이 분산되어 있다는 느낌을 받았기 때문입니다.** 따라서, 이를 모두 최상위 컴포넌트로 옮길지 고민했습니다.

그러나 useQuantityControl이 낙관적 업데이트를 수행하고 있기 때문에, 호출될 때 cartItem을 인자로 받아야 했습니다. 하지만 최상위 컴포넌트인 Cart는 **핸들러 함수를 생성할 시점에 이 핸들러 함수가 어떤 CartItem에 종속될지 알 수 없기 때문에,** 반드시 CartItem 컴포넌트 안에서만 훅을 호출해야 하는 문제가 있었습니다.


따라서 두 가지 방법을 고민했습니다. 

1. CartItem 내부에서 useQuantityControl을 호출해서 핸들러 함수가 사용될 CartItem 정보만 훅에 전달한다. (현상 유지)

2. useQuantityControl 내부의 핸들러 함수가 cartId를 인자로 받도록 하여, cartItems 배열에서 타겟 CartItem을 찾아 처리하도록 한다.

1번의 경우, useQuantityControl가 필요한 데이터만 받지만, **기존의 구조를 유지**하기 때문에 제가 불편함을 느낀 점이 해결되지 않았습니다. 
2번의 경우, 최상위 컴포넌트에서 모든 핸들러 함수를 생성하고 하위 컴포넌트에는 props만 내려주기 때문에 구조가 깔끔해지지만, **useQuantityControl에 cartItems를 통째로 넘겨주어야 했습니다.**


그러나 최종적으로는 유지보수와 데이터 흐름의 일관성을 우선시 하여, 최상위 컴포넌트인 Cart에서 모든 패칭과 상태 변경을 관리하기로 했습니다. 이를 위해 훅에 cartItems 전체를 인자로 넘기고, 핸들러 함수가 cartId로 배열에서 매칭되는 CartItem을 찾아 처리하도록 구조를 변경했습니다.

**[변경 전]**

- CartItem에서 호출

```javascript
  const { decreaseQuantity, increaseQuantity, deleteCartItem } =
    useQuantityControl({
      quantity,
      handleQuantity,
      refetchCartItem: fetchCartItem,
    });
...

  const increaseQuantity = async (cartId: number) => {
    const newQuantity = quantity + 1;
    handleQuantity((prev) => prev + 1);

    await mutateData({
      onError: () => {
        handleQuantity((prev) => prev - 1);
      },
    });
  };
```

<br>

**[변경 후]**

- Cart에서 호출

```javascript
  const { decreaseQuantity, increaseQuantity, deleteCartItem } =
    useQuantityControl({
      cartItems,
      updateCartItem,
      refetchCartItem: fetchCartItem,
    });
...

  const updateQuantity = (cartId: number, quantity: number) => {
    const { product } = findCartItemById(cartId, cartItems);
    const newCartItem = { id: cartId, product, quantity };
    updateCartItem(cartId, newCartItem);
  };

  const increaseQuantity = async (cartId: number) => {
    const { quantity } = findCartItemById(cartId, cartItems);
    const newQuantity = quantity + 1;
    updateQuantity(cartId, newQuantity);

    await fetchData({
      onError: () => {
        updateQuantity(cartId, quantity);
      },
    });
  };
```

이러한 구조 변경에 대해, 쵸파는 어떻게 생각하시는지 궁금합니다..! 다소 과도하게 정보를 넘겨주고 있다고 해도, 최상위 컴포넌트에서 훅을 호출해 핸들러를 생성하는 구조와, 각 CartItem 컴포넌트에서 필요한 정보만 받아 핸들러를 만드는 구조를 결국 프로그래머의 선택의 문제로 볼 수 있는 지도 궁금합니다.


아래는 관련 커밋입니다!

[refactor: cartItems를 변경하는 함수를 최상단 컴포넌트에서 관리하도록 수정](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/efbe16c9ca90fc712feaa0431b4034a2624a5629)

[refactor: cartItem 내부에서 관리하는 핸들러 로직을 최상위 컴포넌트(Cart)로 이동](https://github.com/woowacourse/react-shopping-cart/pull/354/commits/0b5b3c8fd8ae720fa384e18c62f4b18e9098f29c)

---

감사합니다!

bassyu reviewed (approved):
안녕하세요 제나!

코멘트 고민 해주시고 리팩터링 잘 해주셨네요~!

> 상태 관련 로직이 분산되어 있다는 느낌

`useQuantityControl`이라는 관심사를 나눠주는 것 자체는 중요해 보이는데요,
이것을 어디에서 먼저 쓸지는 둘 다 좋아 보여서 선택하기 어렵네요~!

다시 코멘트 남긴 부분은 2단계에 마저 고민해봐도 좋을 것 같습니다~!
2단계도 파이팅입니다! 💪 💪
